File: Kirigami/scripts/plot_k9b.py
#!/usr/bin/env python3
"""K9b figures.

  k9b_gallery.png -- >= 20 deployed random graphs, each shown closed (theta = 0) and at
                     theta = Theta_max / 2, from the JSON dumped by kill_k9b.
  k9b_hist.png    -- the certified eps_max distribution of K9b against the baselines,
                     every one of which is identically 0.

Run with: arch -arm64 /usr/local/bin/python3 code/scripts/plot_k9b.py [results/kill/k9b]
"""
import csv
import glob
import json
import os
import sys

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.patches import Polygon
from matplotlib.collections import PatchCollection

D = sys.argv[1] if len(sys.argv) > 1 else "results/kill/k9b"
G = os.path.join(D, "gallery")

CLOSED = "#BDD2FD"
OPEN = "#5B8FF9"


def load_index():
    rows = []
    for p in sorted(glob.glob(os.path.join(G, "index_*.json"))):
        rows.extend(json.load(open(p)))
    # one entry per graph: the sigma with the larger Theta_max (the honest best-of-2)
    best = {}
    for r in rows:
        k = (r["kind"], r["id"])
        if k not in best or r["theta_max"] > best[k]["theta_max"]:
            best[k] = r
    out = sorted(best.values(), key=lambda r: -r["theta_max"])
    return out, rows


def draw(ax, path, color, title):
    d = json.load(open(path))
    V = d["vertices"]
    patches = [Polygon([V[k] for k in f], closed=True) for f in d["faces"]]
    ax.add_collection(PatchCollection(patches, facecolor=color, edgecolor="0.3",
                                      linewidths=0.15))
    xs = [p[0] for p in V]
    ys = [p[1] for p in V]
    mx = 0.04 * max(max(xs) - min(xs), max(ys) - min(ys), 1e-9)
    ax.set_xlim(min(xs) - mx, max(xs) + mx)
    ax.set_ylim(min(ys) - mx, max(ys) + mx)
    ax.set_aspect("equal")
    ax.axis("off")
    if title:
        ax.set_title(title, fontsize=6.5, pad=1.5)


def gallery(n_want=20):
    entries, allrows = load_index()
    entries = [e for e in entries
               if os.path.exists(os.path.join(G, e["tag"] + "_closed.json"))
               and os.path.exists(os.path.join(G, e["tag"] + "_half.json"))]
    if not entries:
        print("gallery: nothing dumped")
        return 0
    entries = entries[:max(n_want, min(len(entries), 24))]
    ncol = 8  # 4 designs per row, each as a (closed, half) pair
    nrow = (len(entries) + 3) // 4
    fig, axes = plt.subplots(nrow, ncol, figsize=(1.55 * ncol, 1.75 * nrow))
    axes = axes.reshape(nrow, ncol)
    for a in axes.ravel():
        a.axis("off")
    for i, e in enumerate(entries):
        r, cpair = divmod(i, 4)
        a0 = axes[r][2 * cpair]
        a1 = axes[r][2 * cpair + 1]
        draw(a0, os.path.join(G, e["tag"] + "_closed.json"), CLOSED,
             "%s %d %s\nF=%d  closed" % (e["kind"][:4], e["id"],
                                         "mc" if e["sigma"].endswith("mc") else "def",
                                         e["F"]))
        draw(a1, os.path.join(G, e["tag"] + "_half.json"), OPEN,
             r"$\theta=\Theta_{max}/2$" + "\n%.3f rad" % (0.5 * e["theta_max"]))
    fig.suptitle("K9b: %d random graphs that deploy under the convexity + split-inward "
                 "constrained embedding\n(every baseline on the same population is "
                 "$\\Theta_{max}=0$ on all 400 designs)" % len(entries), fontsize=9)
    fig.tight_layout(rect=[0, 0, 1, 0.955])
    out = os.path.join(D, "k9b_gallery.png")
    fig.savefig(out, dpi=190)
    plt.close(fig)
    print("wrote %s with %d designs" % (out, len(entries)))
    return len(entries)


def union_positives(rows):
    """Per design, the better of K9's solver configuration and K9b's sweep. Both are
    exact-verified positives under the same certificate on the same 400 designs, so the
    per-design maximum is a legitimate multi-start count -- and it is reported next to
    the two runs, never in place of either."""
    path = "results/kill/k9/k9.csv"
    if not os.path.exists(path):
        return sum(1 for r in rows if float(r["theta_ref9"]) > 1e-9)
    K = {(r["kind"], r["id"], r["sigma"]): r for r in csv.DictReader(open(path))}
    n = 0
    for r in rows:
        k = (r["kind"], r["id"], r["sigma"])
        best = float(r["theta_exact"])
        if k in K:
            best = max(best, float(K[k]["b_theta_exact"]))
        if best > 1e-9:
            n += 1
    return n


def hist():
    path = os.path.join(D, "k9b.csv")
    if not os.path.exists(path):
        print("no k9b.csv")
        return
    rows = list(csv.DictReader(open(path)))
    eps = [float(r["eps_max"]) for r in rows if float(r["eps_max"]) > 0]
    fig, ax = plt.subplots(1, 2, figsize=(10.0, 3.8))

    ax[0].hist(eps, bins=24, color="#5B8FF9", edgecolor="0.25", linewidth=0.5)
    ax[0].axvline(0.1, color="#E8684A", ls="--", lw=1.2)
    ax[0].text(0.105, ax[0].get_ylim()[1] * 0.92, "0.1 rad target", color="#E8684A",
               fontsize=8)
    ax[0].set_xlabel(r"certified $\varepsilon_{max}$ (rad)")
    ax[0].set_ylabel("designs")
    ax[0].set_title("K9b: %d certified designs of %d" % (len(eps), len(rows)), fontsize=9)

    # Every baseline on this exact population is identically zero.
    names = ["Eq. (6)\n(K1a)", r"$\sigma_{mc}$" + "\n(K5)", r"$\sigma_{def}$" + "\n(K5)",
             "0+ repair\n(K6)", "convexity\nonly (K9a)", "K9 (b)", "K9b\nsweep",
             "best of\nboth"]
    k9 = 36
    # A positive needs BOTH the exact scan and the referee: delaunay 40 sigma_mc has
    # bisection(1e-9) = 2.07e-7 rad, below the bisection's own grid resolution, with the
    # exact scan at 0. It is bisection noise, not a design that deploys.
    k9b = sum(1 for r in rows
              if float(r["theta_exact"]) > 1e-9 and float(r["theta_ref9"]) > 1e-9)
    union = union_positives(rows)
    vals = [0, 0, 0, 0, 0, k9, k9b, union]
    cols = ["0.7"] * 5 + ["#9BB7F0", "#9BB7F0", "#5B8FF9"]
    ax[1].bar(range(len(vals)), vals, color=cols, edgecolor="0.25", linewidth=0.5)
    ax[1].axhline(40, color="#E8684A", ls="--", lw=1.2)
    ax[1].text(0.05, 41, "PASS bar 40 / 400", color="#E8684A", fontsize=8)
    ax[1].set_xticks(range(len(names)))
    ax[1].set_xticklabels(names, fontsize=7)
    ax[1].set_ylabel(r"designs with refereed $\Theta_{max} > 0$")
    ax[1].set_title("of 400 designs (200 graphs x 2 sigma)", fontsize=9)
    for i, v in enumerate(vals):
        ax[1].text(i, v + 0.8, str(v), ha="center", fontsize=8)

    fig.tight_layout()
    out = os.path.join(D, "k9b_hist.png")
    fig.savefig(out, dpi=180)
    plt.close(fig)
    print("wrote %s" % out)


if __name__ == "__main__":
    gallery()
    hist()

File: Kirigami/scripts/plot_k9b.jl
#!/usr/bin/env julia
# K9b figures.
#   k9b_gallery.png -- >= 20 deployed random graphs, each shown closed (theta = 0) and at
#                      theta = Theta_max / 2, from the JSON dumped by kill_k9b.
#   k9b_hist.png    -- the certified eps_max distribution of K9b against the baselines.
# Run: julia --project=Kirigami/scripts Kirigami/scripts/plot_k9b.jl [results/kill/k9b]
include(joinpath(@__DIR__, "plot_common.jl"))

D = length(ARGS) >= 1 ? ARGS[1] : "results/kill/k9b"
G = joinpath(D, "gallery")

const CLOSED = "#BDD2FD"
const OPEN = "#5B8FF9"

function load_index()
    all = Any[]
    for p in sort(filter(x -> startswith(basename(x), "index_") && endswith(x, ".json"),
                         readdir(G; join = true)))
        append!(all, JSON.parsefile(p))
    end
    # one entry per graph: the sigma with the larger Theta_max (the honest best-of-2)
    best = Dict{Tuple{String,Int},Any}()
    for r in all
        k = (r["kind"], Int(r["id"]))
        if !haskey(best, k) || r["theta_max"] > best[k]["theta_max"]
            best[k] = r
        end
    end
    return sort(collect(values(best)); by = r -> -r["theta_max"]), all
end

function draw!(ax, path, color, title)
    d = JSON.parsefile(path)
    V = d["vertices"]
    draw_faces!(ax, V, d["faces"]; color, strokecolor = GREY(0.3), strokewidth = 0.15)
    frame_axis!(ax, V; margin = 0.04)
    ax.title = title
end

function gallery(n_want = 20)
    entries, _ = load_index()
    entries = [e for e in entries if isfile(joinpath(G, e["tag"] * "_closed.json")) &&
                                     isfile(joinpath(G, e["tag"] * "_half.json"))]
    if isempty(entries)
        println("gallery: nothing dumped")
        return 0
    end
    entries = entries[1:min(length(entries), max(n_want, min(length(entries), 24)))]
    ncol = 8  # 4 designs per row, each as a (closed, half) pair
    nrow = (length(entries) + 3) ÷ 4
    fig = Figure(size = (155 * ncol, 175 * nrow + 60))
    for (i, e) in enumerate(entries)
        r, cpair = divrem(i - 1, 4)
        a0 = Axis(fig[r + 1, 2cpair + 1]; titlesize = 8)
        a1 = Axis(fig[r + 1, 2cpair + 2]; titlesize = 8)
        draw!(a0, joinpath(G, e["tag"] * "_closed.json"), CLOSED,
              @sprintf("%s %d %s\nF=%d  closed", first(e["kind"], 4), e["id"],
                       endswith(e["sigma"], "mc") ? "mc" : "def", e["F"]))
        draw!(a1, joinpath(G, e["tag"] * "_half.json"), OPEN,
              @sprintf("θ = Θmax/2\n%.3f rad", 0.5 * e["theta_max"]))
    end
    Label(fig[0, 1:ncol],
          "K9b: $(length(entries)) random graphs that deploy under the convexity + split-inward " *
          "constrained embedding\n(every baseline on the same population is Θmax = 0 on all 400 designs)";
          fontsize = 11)
    out = outpath(joinpath(D, "k9b_gallery.png"))
    savefig(fig, out)
    println("wrote $out with $(length(entries)) designs")
    return length(entries)
end

# Per design, the better of K9's solver configuration and K9b's sweep. Both are exact-verified
# positives under the same certificate on the same 400 designs, so the per-design maximum is a
# legitimate multi-start count -- reported next to the two runs, never in place of either.
function union_positives(R)
    path = "results/kill/k9/k9.csv"
    isfile(path) || return count(r -> fnum(r, "theta_ref9") > 1e-9, R)
    K = Dict((r["kind"], r["id"], r["sigma"]) => r for r in rows(path))
    n = 0
    for r in R
        k = (r["kind"], r["id"], r["sigma"])
        best = fnum(r, "theta_exact")
        haskey(K, k) && (best = max(best, fnum(K[k], "b_theta_exact")))
        best > 1e-9 && (n += 1)
    end
    return n
end

function hist()
    path = joinpath(D, "k9b.csv")
    isfile(path) || (println("no k9b.csv"); return)
    R = rows(path)
    eps = [fnum(r, "eps_max") for r in R if fnum(r, "eps_max") > 0]
    fig = Figure(size = (1000, 380))
    ax0 = Axis(fig[1, 1]; xlabel = L"certified $\varepsilon_{max}$ (rad)", ylabel = "designs",
               title = "K9b: $(length(eps)) certified designs of $(length(R))", titlesize = 11)
    hist_bars!(ax0, eps, linbins(eps, 24); color = "#5B8FF9", strokecolor = GREY(0.25),
               strokewidth = 0.5)
    vlines!(ax0, [0.1]; color = "#E8684A", linestyle = :dash, linewidth = 1.2)
    ymax = maximum(bincounts(eps, linbins(eps, 24)); init = 1)
    text!(ax0, 0.105, ymax * 0.92; text = "0.1 rad target", color = "#E8684A", fontsize = 10)

    # Every baseline on this exact population is zero except the K6 repair, read from k6.csv
    # under its best-of-four rule.
    names = ["Eq. (6)\n(K1a)", "σ_mc\n(K5)", "σ_def\n(K5)", "0+ repair\n(K6, best of 4)",
             "convexity\nonly (K9a)", "K9 (b)", "K9b\nsweep", "best of\nboth"]
    k9 = 36
    # A positive needs BOTH the exact scan and the referee: delaunay 40 sigma_mc has
    # bisection(1e-9) = 2.07e-7 rad, below the bisection's own grid resolution, with the exact
    # scan at 0. It is bisection noise, not a design that deploys.
    k9b = count(r -> fnum(r, "theta_exact") > 1e-9 && fnum(r, "theta_ref9") > 1e-9, R)
    union = union_positives(R)
    vals = [0, 0, 0, k6_baseline_deployable(), 0, k9, k9b, union]
    cols = vcat(fill(GREY(0.7), 5), parse.(Makie.Colors.Colorant, ["#9BB7F0", "#9BB7F0", "#5B8FF9"]))
    ax1 = Axis(fig[1, 2]; xticks = (0:length(vals)-1, names), xticklabelsize = 9,
               ylabel = L"designs with refereed $\Theta_{max} > 0$",
               title = "of 400 designs (200 graphs x 2 sigma)", titlesize = 11)
    barplot!(ax1, 0:length(vals)-1, vals; color = cols, strokecolor = GREY(0.25), strokewidth = 0.5)
    hlines!(ax1, [40]; color = "#E8684A", linestyle = :dash, linewidth = 1.2)
    text!(ax1, 0.05, 41; text = "PASS bar 40 / 400", color = "#E8684A", fontsize = 10)
    for (i, v) in enumerate(vals)
        text!(ax1, i - 1, v + 0.8; text = string(v), align = (:center, :bottom), fontsize = 10)
    end
    out = outpath(joinpath(D, "k9b_hist.png"))
    savefig(fig, out)
end

gallery()
hist()
